### PR TITLE
Api baselines

### PR DESCRIPTION
**Pull request by Pilchie**
Resolves part of #24347

### CODE REVIEW

File: src/FileProviders/Manifest.MSBuildTask/src/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+#nullable enable
```
Comment (Pilchie) [inline src/FileProviders/Manifest.MSBuildTask/src/PublicAPI.Unshipped.txt:1]: Another example of a project that has public API, but the analyzer isn't running on.

File: src/Analyzers/Analyzers/src/PublicAPI.Shipped.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+#nullable enable
```
Comment (Pilchie) [inline src/Analyzers/Analyzers/src/PublicAPI.Shipped.txt:1]: @mavasani Regarding this, how much does it matter whether we've actually started annotating these libraries here?

Comment (mavasani) [reply src/Analyzers/Analyzers/src/PublicAPI.Shipped.txt:1]: @jcouv would be a better person to answer this.

**Comment by Pilchie** [comment]
Hmm.  VS had no errors, but the commandline is definitely unhappy 😢

File: src/Analyzers/Analyzers/src/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+#nullable enable
```
[the remark this answers is not in the document]
Comment (dougbu) [reply src/Analyzers/Analyzers/src/PublicAPI.Unshipped.txt:1]: I specifically include only projects with ` '$(IsImplementationProject)' == 'true' ` and that excludes analyzer projects. From a compatibility POV, adding or removing `public` surface in these projects shouldn't matter. So, these files can be removed.

Side note: I also excluded all projects in the src/Tools/ directory (well, anything w/ a relative path that includes "Tools") for similar reasons. Might need to exclude `msbuild` tasks too (for the Razor SDK) but I didn't get there.

File: AspNetCore.sln [hunk 4]
```diff
@@ -1043,7 +1043,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Razor.
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Microsoft.AspNetCore.Razor.Tools", "Microsoft.AspNetCore.Razor.Tools", "{B9704650-5360-416C-9393-FAF707766AA8}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Razor.Tools", "src\Razor\Microsoft.AspNetCore.Razor.Tools\src\rzc.csproj", "{4C84173F-5C1F-49A1-895A-C0CA11DE84B1}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "rzc", "src\Razor\Microsoft.AspNetCore.Razor.Tools\src\rzc.csproj", "{4C84173F-5C1F-49A1-895A-C0CA11DE84B1}"
```
Comment (dougbu) [inline AspNetCore.sln:1046]: Sorry, my bad.

**Comment by dougbu** [comment]
@Pilchie on the failures, it looks like you're hitting a conflict with 00bbb78d3c96 aka @pranavkm's #25275 fix.

**Comment by dougbu** [comment]
Sorry, unintended the close

File: AspNetCore.sln (hunk 4 above)
Comment (Pilchie) [reply AspNetCore.sln:1046]: Meh - VS fixed that part without me doing anything.

File: src/FileProviders/Manifest.MSBuildTask/src/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (dougbu) [reply src/FileProviders/Manifest.MSBuildTask/src/PublicAPI.Unshipped.txt:1]: Project file contains `<IsImplementationProject>false</IsImplementationProject>` so it's excluded from the analysis

File: src/HealthChecks/HealthChecks/src/DependencyInjection/HealthChecksBuilderAddCheckExtensions.cs [hunk 5]
```diff
@@ -48,7 +48,9 @@ public static class HealthChecksBuilderAddCheckExtensions
         /// <param name="tags">A list of tags that can be used to filter health checks.</param>
         /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
         /// <returns>The <see cref="IHealthChecksBuilder"/>.</returns>
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
```
Comment (dougbu) [inline src/HealthChecks/HealthChecks/src/DependencyInjection/HealthChecksBuilderAddCheckExtensions.cs:51]: @sharwell recommended using the following in his review of #24188:
``` msbuild
[SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Required to maintain compatibility")]
```
See https://github.com/dotnet/aspnetcore/pull/24188#discussion_r458957296 but this also reduces the noise to a single line😺

File: src/HealthChecks/HealthChecks/src/DependencyInjection/HealthChecksBuilderAddCheckExtensions.cs [hunk 6]
```diff
@@ -121,7 +123,9 @@ public static class HealthChecksBuilderAddCheckExtensions
         /// with any lifetime it will be used. Otherwise an instance of type <typeparamref name="T"/> will be constructed with
         /// access to services from the dependency injection container.
         /// </remarks>
+#pragma warning disable RS0027 // Public API with optional parameter(s) should have the most parameters amongst its public overloads.
```
Comment (dougbu) [inline src/HealthChecks/HealthChecks/src/DependencyInjection/HealthChecksBuilderAddCheckExtensions.cs:126]: Nice, we don't have any RS0027 suppressions yet🥇

I'd use the following here:
``` c#
[SuppressMessage("ApiDesign", "RS0027:Public API with optional parameter(s) should have the most parameters amongst its public overloads.", Justification = "Required to maintain compatibility")]
```

File: src/Analyzers/Analyzers/src/PublicAPI.Unshipped.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (dougbu) [reply src/Analyzers/Analyzers/src/PublicAPI.Unshipped.txt:1]: BTW I wasn't sure whether the analyzers folder contained anything w/ important public API e.g. a base package when I wrote up #24347. Sorry.

File: src/Analyzers/Analyzers/src/PublicAPI.Shipped.txt (hunk 2 above)
Comment (dougbu) [reply src/Analyzers/Analyzers/src/PublicAPI.Shipped.txt:1]: Based on the lack of errors about this, I suggest we include `#nullable enable` everywhere. Makes life simpler.

**Review by dougbu — CHANGES_REQUESTED** [review]
Let's consistently use attributes for the RS00XY suppressions

**Comment by Pilchie** [comment]
Hmm.  I'm going to temporarily remove the PublicAPI files from the AzureAD.UI and AzureADB2C.UI projects. Seems like there are a couple of problems with using them in razor projects:

1. The analyzer gets passed to both compilation phases, so if the razor generated members are present, it complains during the second compile, but if they *are* present it complains during the first compile 😦 .
2. The fix in VS doesn't actually generate anything for the razor generated members.

@mavasani any suggestions on how do support this?

@chsienki - this is similar to the source generator issue with razor compilation.  Any ideas?

@pranavkm / @jaredpar - FYI.

File: src/HealthChecks/HealthChecks/src/HealthCheckService.cs [hunk 7]
```diff
@@ -38,6 +39,7 @@ public abstract class HealthCheckService
         /// A <see cref="Task{T}"/> which will complete when all the health checks have been run,
         /// yielding a <see cref="HealthReport"/> containing the results.
         /// </returns>
+        [SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Required to maintain compatibility")]
         public Task<HealthReport> CheckHealthAsync(CancellationToken cancellationToken = default)
```
Comment (dougbu) [inline src/HealthChecks/HealthChecks/src/HealthCheckService.cs:43]: nit: I wish the analyzer would suppress itself when the only optional parameter is a `CancellationToken`. The "overload for `int`, `float`, &hellip; cases e.g. `WriteAsync(...)` methods and similar are causing a lot of suppression noise.

@sharwell @mavasani @jcouv❔

**Review by dougbu — APPROVED** [review]

**Comment by Pilchie** [comment]
I'm going to go ahead and approve my own PR.

**Comment by dougbu** [comment]
Didn't I approve it❔ And, ugh, you're hitting the overlapping PDB problem here too☹️

**Comment by Pilchie** [comment]
I mean ask-mode approval ;)

**Pull request merged**